File: transports/elixir/lib/csilgen/transport/status.ex
defmodule Csilgen.Transport.Status do
  @moduledoc """
  Transport-level status codes (conventions doc §4). Distinct from application
  errors, which ride inside the payload as a declared `/ ErrorType` arm.

  A status is carried on the wire as a plain integer code, so it is kept as an
  `integer()` in the envelope structs (equality is trivially by code, including
  host-defined extension codes ≥ 64 and unknown codes).
  """

  @ok 0
  @malformed_envelope 1
  @unknown_service_or_op 2
  @unauthenticated 3
  @forbidden 4
  @version_unsupported 5
  @internal 6
  @unavailable 7
  @deadline_exceeded 8

  @doc "Status code constant for `ok`."
  def ok, do: @ok
  def malformed_envelope, do: @malformed_envelope
  def unknown_service_or_op, do: @unknown_service_or_op
  def unauthenticated, do: @unauthenticated
  def forbidden, do: @forbidden
  def version_unsupported, do: @version_unsupported
  def internal, do: @internal
  def unavailable, do: @unavailable
  def deadline_exceeded, do: @deadline_exceeded

  @doc "Whether the status indicates a typed reply is present."
  @spec ok?(integer()) :: boolean()
  def ok?(@ok), do: true
  def ok?(_), do: false

  @doc "The registry name for a status code, or \"other\" for codes outside it."
  @spec name(integer()) :: String.t()
  def name(@ok), do: "ok"
  def name(@malformed_envelope), do: "malformed-envelope"
  def name(@unknown_service_or_op), do: "unknown-service-or-op"
  def name(@unauthenticated), do: "unauthenticated"
  def name(@forbidden), do: "forbidden"
  def name(@version_unsupported), do: "version-unsupported"
  def name(@internal), do: "internal"
  def name(@unavailable), do: "unavailable"
  def name(@deadline_exceeded), do: "deadline-exceeded"
  def name(_), do: "other"
end

File: transports/java/src/main/java/community/catalyst/csilgen/transport/Status.java
// A transport-level status. It is distinct from application errors, which ride inside
// the payload as a declared `/ ErrorType` arm (conventions doc §4). Equality is by the
// underlying code, so host-defined extension codes (>= 64) and unknown codes compare
// correctly.
package community.catalyst.csilgen.transport;

public record Status(long code) {
    public static final Status OK = new Status(0);
    public static final Status MALFORMED_ENVELOPE = new Status(1);
    public static final Status UNKNOWN_SERVICE_OR_OP = new Status(2);
    public static final Status UNAUTHENTICATED = new Status(3);
    public static final Status FORBIDDEN = new Status(4);
    public static final Status VERSION_UNSUPPORTED = new Status(5);
    public static final Status INTERNAL = new Status(6);
    public static final Status UNAVAILABLE = new Status(7);
    public static final Status DEADLINE_EXCEEDED = new Status(8);

    /** Maps a wire code onto a Status, preserving host-defined and unknown codes. */
    public static Status fromCode(long code) {
        return new Status(code);
    }

    /** Whether the status indicates a typed reply is present. */
    public boolean isOk() {
        return code == 0;
    }

    /** The registry name for the status, or "other" for codes outside the registry. */
    public String name() {
        return switch ((int) code) {
            case 0 -> "ok";
            case 1 -> "malformed-envelope";
            case 2 -> "unknown-service-or-op";
            case 3 -> "unauthenticated";
            case 4 -> "forbidden";
            case 5 -> "version-unsupported";
            case 6 -> "internal";
            case 7 -> "unavailable";
            case 8 -> "deadline-exceeded";
            default -> "other";
        };
    }
}
